marimo/_plugins/stateless/mpl.py

```diff
@@ -0,0 +1,344 @@
+            # the FigureManager.
+            manager = self.application.manager  # type: ignore[attr-defined]
+            manager.remove_web_socket(self)
+
+        def on_message(self, message: Any) -> None:
+            # The 'supports_binary' message is relevant to the
+            # websocket itself.  The other messages get passed along
+            # to matplotlib as-is.
+
+            # Every message has a "type" and a "figure_id".
+            message = json.loads(message)
+            if message["type"] == "supports_binary":
+                self.supports_binary = message["value"]
+            else:
+                manager = self.application.manager  # type: ignore[attr-defined] # noqa: E501
+                manager.handle_json(message)
+
+        def send_json(self, content: str) -> None:
+            self.write_message(json.dumps(content))
+
+        def send_binary(self, blob: Any) -> None:
+            if self.supports_binary:
+                self.write_message(blob, binary=True)
+            else:
+                data_uri = "data:image/png;base64," + blob.encode(
+                    "base64"
+                ).replace("\n", "")
+                self.write_message(data_uri)
+
+    def __init__(self, figure: Any) -> None:
+        import matplotlib as mpl  # type: ignore[import]
+        from matplotlib.backends.backend_webagg import (
+            FigureManagerWebAgg,
+            new_figure_manager_given_figure,
+        )
+
+        self.figure = figure
+        self.manager = new_figure_manager_given_figure(id(figure), figure)
+
+        super().__init__(
+            [
+                # Static files for the CSS and JS
+                (
+                    r"/_static/(.*)",
+                    tornado.web.StaticFileHandler,
+                    {"path": FigureManagerWebAgg.get_static_file_path()},
+                ),
+                # Static images for the toolbar
+                (
+                    r"/_images/(.*)",
+                    tornado.web.StaticFileHandler,
+                    {"path": Path(mpl.get_data_path(), "images")},
+                ),
+                # The page that contains all of the pieces
+                ("/", self.MainPage),
+                ("/mpl.js", self.MplJs),
+                # Sends images and events to the browser, and receives
+                # events from the browser
+                ("/ws", self.WebSocket),
+                # Handles the downloading (i.e., saving) of static images
+                (r"/download.([a-z0-9.]+)", self.Download),
+            ]
+        )
+
+
+@dataclass
+class CleanupHandle:
+    """Handle to shutdown a figure server."""
+
+    shutdown_event: Optional[asyncio.Event] = None
+
+
+class InteractiveMplRegistry:
+    """Registry of figures created by each cell.
+
+    Allows the kernel to tear down all figures for a given cell.
+    """
+
+    _registry: dict[CellId_t, list[CleanupHandle]] = {}
```

Comment:
should this be a set of CleanupHandle to prevent duplicates? if that is even possible

Reply:
It's tricky because you'd want set membership to depend on the id of the shutdown event, but that might not exist when the handle is first registered. `CleanupHandle.shutdown_event` starts off as `None` and is later re-assigned to an `Event` in the server thread, for compatibility with Python 3.8 and 3.9 (creating the event in the main thread lead to an error). 

Right now a duplicate Event won't ever be registered. I agree for defensibility a set would be nice ... but I think in this case it's not worth the extra effort (i.e. synchronizing to wait until the shutdown event is created before registering).